README.md
# Salesforce Commerce Cloud log tail

> Remote tail Salesforce Commerce Cloud logs via webdav. Allow to monitor more logs at once, merging the content of all the files in a single stream. Reports either to console or a FluentD collector.

## Features

-   Authentication using API Client _(recommended)_ OR Business Manager _(deprecated)_
-   Interactive prompt for logs selection OR selection of logs by config file.
-   Supports configuration of multiple instances OR standard dw.json config file
-   Outputs to console OR FluentD collector
-   Multiple log tailing, with merging of log entries
-   In Console mode:
    -   Sorts log entries by timestamp
    -   Colors output based on log levels
    -   Converts log timestamp to local timezone

## Installation

```bash
$ npm i -g cctail
```

## Requirements

-   Node >= 10

## Configuration

Requires one of the following configuration files:

-   a `log.conf.json` file with multiple environments configured. This may be used if you want to easily switch between multiple instances
-   a standard `dw.json` file, tipically pointing to your working sandbox.

`cctail` requires a correctly configured API client id/secret OR Business Manager username/password for accessing logs via webdav. **API client authentication is recommended, because it is faster after the initial authorization, and Business Manager authentication to WebDAV has been _deprecated_ by SalesForce.**

### Optional Configurations

-   `"profiles"`: - Standard log types: `analytics`, `api`, `console`, `customdebug`, `customerror`, `customfatal`, `custominfo`, `customwarn`, `dbinit-sql`, `debug`, `deprecation`, `error`, `fatal`, `info`, `jobs`, `migration`, `performance`, `quota`, `sql`, `staging`, `sysevent`, `syslog`, `warn`
    -   `"log_types": ["log", "types", "array"]` _(default: all log types)_ - In non-interactive mode, defining this will limit the log types that cctail collects to this list.
        -   Standard log types: `analytics`, `api`, `codeprofiler`, `console`, `customdebug`, `customerror`, `customfatal`, `custominfo`, `customwarn`, `dbinit-sql`, `debug`, `deprecation`, `error`, `fatal`, `info`, `jobs`, `migration`, `performance`, `quota`, `sql`, `staging`, `sysevent`, `syslog`, `warn`
    -   `"polling_interval": nnn` _(default: `3`)_ - Frequency (seconds) with which cctail will poll the logs.
        -   If you are using non-interactive mode to pipe the logs elsewhere (i.e. FluentD), a longer interval is recommended (i.e. 30 or 60).
    -   `"log_source"`_(default:`operational`)_: flag for routing from where to read logs. Available 3 parameters `all`, `security`, `operational`.
    -   `"refresh_loglist_interval": nnn` _(default: `600`)_ - In non-interactive mode, this is the frequency (seconds) in which cctail will check the WebDAV server for new logs that match your `log_types` criteria.
-   `"interactive": true|false` _(default: `true`)_ - Interactive mode asks which logs you will want to tail. If `false`, cctail will tail all of today's logs by default.

#### FluentD

**NOTE:** All configurations for fluentD are _optional_, except `enabled` must be set to `true` if you want to use it.

-   `fluent`:
    -   `"enabled": true|false` _(default: `false`)_ - If enabled, logs will be directed to the Fluent collector.
    -   `"host": "fluentd.yourco.com"` _(default: `localhost`)_ - FluentD collector host
    -   `"port": nnn` _(default: `24224`)_ - FluentD collector port
    -   `"reconnect_interval": nnn` _(default: `600`)_ - If the collector can't be reached, cctail will try to reconnect again in `nnn` seconds.
    -   `"timeout": nnn` _(default: `3`)_ - Timeout to connect to FluentD collector
    -   `"tag_prefix": "your_tag_prefix"` _(default: `sfcc`)_ - All logs sent to FluentD will have this prefix, followed by the log type (i.e. "sfcc.customerror").

### Sample configuration files

Sample dw.json:

```json
{
    "hostname": "dev01-mysandbox.demandware.net",
    "client_id": "a12464ae-b484-4b90-4dfe-17e20844e9f0",
    "client_secret": "mysupersecretpassword",
    "log_source": "all | security | operational"
}
```

Sample log.conf.json:

```json
{
    "profiles": {
        "dev01-api-client-example": {
            "hostname": "dev01-mysandbox.demandware.net",
            "client_id": "a12345ae-b678-9b01-2dfe-34e56789e0f1",
            "client_secret": "mysupersecretsecret",
            "polling_interval": 30,
            "refresh_loglist_interval": 900
        },
        "dev02-bm-example": {
            "hostname": "dev02-mysandbox.demandware.net",
            "username": "user@yourco.com",
            "password": "mysupersecretpassword",
            "log_types": ["customerror", "customwarn", "error", "jobs", "warn"],
            "polling_interval": 60
        }
    },
    "interactive": false,
    "fluent": {
        "enabled": true
    }
}
```

If multiple instances are configured, you may directly pass the name of the instance for skipping the interactive selection prompt, e.g.:

```bash
$ cctail dev02
```

### API client configuration

The API client id must be created in the account.demandware.com console. Before being able to use `cctail` you must grant the required permissions to that client id for accessing the logs folder through WebDAV in any target SFCC instance.

To do so, access Business Manager and add the following to Administration -> Organization -> WebDAV Client Permissions, replacing the client_id value with your client id. **Note:** you may need to merge these settings with existing ones.

```json
{
    "clients": [
        {
            "client_id": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
            "permissions": [
                {
                    "path": "/logs",
                    "operations": ["read_write"]
                }
            ]
        }
    ]
}
```

## Usage

```bash
$ cctail
```

Run `cctail` in a folder containing either a log.conf-json or dw.json config file.
The tool will display the list of available logs in order to let you interactively select the ones you want to monitor.

## Credits

This repo is based on a great work that was done by [OpenMindLab](https://github.com/openmindlab/cctail).
We added some more capabilities to better support our customers’ use cases.

## License

Copyright (c) 2019 openmind

Released under the MIT license.


CHANGELOG.md
## Change Log

### v1.6.0 (2022/05/11)

-   Support security, multiple(security+site) routes to fetch logs.
<details>
  <summary markdown="span"> Expand to check old versions </summary>

### v1.5.1 (2021/01/08)

-   Support for environment variables for use in profiles.

### v1.5.0 (2021/01/06)

-   Improved stability, re-authentication is more reliable now.

### v1.4.0 (2020/09/18)

-   Support for FluentD as a log collector
-   Headless mode - tail logs based on config file
-   Added Business Manager as an authentication method
-   Standardized console.log color outputs & debug logging

### v1.3.2 (2020/08/27)

-   Update dependencies
-   Fix parsing of numeric arguments (when using something like 001 as an environment name)

### v1.3.0 (2020/07/27)

-   Improved retry in case of expired auth.

### v1.2.0 (2020/04/26)

-   Update dependencies, migrated to typescript.

### v1.1.3 (2019/11/29)

-   Handling of PIG instances behind cloudflare, even if we are not getting content-length headers here.

### v1.1.2 (2019/08/22)

-   Better sorting of same-second log entries.

### v1.1.1 (2019/08/15)

-   Brighter, more readable log colors.

### v1.1.0 (2019/08/04)

-   Remove dependency on sfcc-ci for authentication. No more salesforce private deps, feeling a bit more free and snappy now.
-   Automatically refresh access token for your long tailing sessions
-   Better handling of hanging connections, now they will timeout sooner so we can retry without loosing too much time.

### v1.0.0 (2019/07/20)

-   First public release


src/lib/types.ts
import moment from 'moment';

export interface DwJson {
	hostname: string,
	client_id?: string,
	log_source?: 'operational' | 'security' | 'all',
	client_secret?: string,
	username?: string,
	password?: string,
	log_types?: string[]
	polling_interval?: number,
	refresh_loglist_interval?: number,
	token?: string,
	token_type?: string,
	token_expiry?: moment.Moment
}

export interface LogFile {
	log: string,
	size_string: string,
	date: moment.Moment,
	size?: number,
	debug: boolean,
	logSource: string
}

export interface LogLine {
	message: string,
	level: string,
	timestamp: moment.Moment,
	logfile?: string,
	logSource?: string

}

export interface LogConfig {
	profiles: Profiles,
	fluent?: FluentConfig,
	interactive?: boolean,
}

export interface Profiles {
	[name: string]: DwJson
}

export interface FluentConfig {
	enabled: boolean,
	host?: string,
	port?: number,
	reconnect_interval?: number,
	timeout?: number,
	tag_prefix?: string
}


src/lib/logfluent.ts
import { LogLine, FluentConfig } from './types';
import logger from './logger';
const fluent = require('fluent-logger');

class LogFluent implements FluentConfig {
	enabled: boolean;
	host: string;
	port: number;
	reconnect_interval: number;
	timeout: number;
	tag_prefix: string;

	constructor(fConfig: FluentConfig) {
		this.enabled = fConfig.enabled;
		this.host = fConfig.host || 'localhost';
		this.port = fConfig.port || 24224;
		this.reconnect_interval = fConfig.reconnect_interval || 600;
		this.timeout = fConfig.timeout || 3.0;
		this.tag_prefix = fConfig.tag_prefix || 'sfcc';

		fluent.configure(this.tag_prefix, {
			host: this.host,
			port: this.port,
			timeout: this.timeout,
			reconnectInterval: this.reconnect_interval * 1000
		});
	}

	output(hostname: string, logs: LogLine[], printnots: boolean, debug: boolean, logSource?: string) {
		logger.log(logger.debug, `Sending ${logs.length} new log records to FluentD this interval.`, debug);

		for (let j = 0; j < logs.length; j++) {
			let log = logs[j];

			if (!printnots && !log.timestamp) {
				// eslint-disable-next-line no-continue
				continue;
			}

			try {
				fluent.emit(log.logfile.substr(0, log.logfile.indexOf('-')), {
					logfile: log.logfile,
					level: log.level,
					message: log.message.trim(),
					hostname: hostname,
					logSource: log.logSource
				}, fluent.EventTime.fromTimestamp(log.timestamp.valueOf()));
			} catch (err) {
				logger.log('error', 'Send to FluentD failed with error: ' + err);
			}
		}
	}
}

export default LogFluent;


src/lib/logparser.ts
import moment from 'moment';
import { LogFile, LogLine } from './types';
import logger from './logger'

const logparser = {
	process: async function (files: Promise<[LogFile, string]>[]): Promise<LogLine[]> {
		return Promise.all(files).then((values) => {
			return values.map((data) => {
				if (data[0].log.endsWith(".csv"))
					return this.parseCsv(data);
				else
					return this.parseLog(data);
			}).reduce((a, b) => {
				if (a.length === 0)
					return b;
				return a.concat(b);
			});
		});
	},

	parseCsv: function (logdata: [LogFile, string]): LogLine[] {
		let logfile = logdata[0].log.replace("/", "-");
		let logSource = logdata[0].logSource;
		let timestamp = logdata[0].date;
		let data = logdata[1];
		let linesobj: LogLine[] = [];
		let regexp = new RegExp((/([^\n]+)/g));
		let mmatch;

		while (mmatch = regexp.exec(data)) {
			let start = mmatch.index;
			let end = start + mmatch[0].length;
			// Start > 0 guarantees we skip the first line, which is the CSV header.
			if (start > 0 && end > 0) {
				linesobj.push({
					message: data.substring(start, end),
					timestamp: timestamp,
					level: logger.profile,
					logfile: logfile,
					logSource
				});
			}
		}
		return linesobj;
	},

	parseLog: function (logdata: [LogFile, string]): LogLine[] {
		let logfile = logdata[0].log;
		let logSource = logdata[0].logSource;
		let data = logdata[1];
		let linesobj: LogLine[] = [];
		let regexp = new RegExp((/\[([.0-9 .:-]*) GMT\] (DEBUG|INFO|WARN|ERROR|FATAL)? ?(.*)/g));
		regexp.lastIndex = 0;
		let lastmatchend = 0;
		let mmatch;
		// eslint-disable-next-line no-cond-assign
		while (mmatch = regexp.exec(data)) {
			let start = mmatch.index;
			let end = start + mmatch[0].length;
			if (start > 0 && linesobj.length === 0) {
				linesobj.push({
					message: data.substring(0, start - 1),
					timestamp: undefined,
					level: undefined,
					logfile: logfile,
					logSource
				});
			}
			else if (start > (lastmatchend + 1)) {
				// append extra lines to the previous log
				linesobj[linesobj.length - 1].message += `${data.substring(lastmatchend, start - 1)}`;
			}
			// 2019-07-15 11:00:07.235
			let timestamp = moment.utc(mmatch[1], "YYYY-MM-DD HH:mm:ss.SSS");
			let level = mmatch[2] || 'INFO';
			let message = mmatch[3];
			lastmatchend = end;
			linesobj.push(
				{
					timestamp,
					level,
					message,
					logfile,
					logSource
				}
			);
		}
		if (linesobj.length > 0 && lastmatchend < data.length) {
			linesobj[linesobj.length - 1].message += `${data.substring(lastmatchend, data.length)}`;
		}
		if (lastmatchend === 0 && data && data.length > 0) { // no match
			linesobj.push({
				message: data,
				timestamp: undefined,
				level: undefined,
				logfile: logfile,
				logSource
			});
		}
		return linesobj;
	}
}

export default logparser;


src/lib/logfetcher.ts
import Axios, { Method, AxiosResponse, AxiosRequestConfig } from 'axios';
import chalk from 'chalk';
import fs from 'fs';
import path from 'path';
import { DwJson, LogFile } from './types';
import logger from './logger'
import moment from 'moment';

const { log } = console;

const ua = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36";
const timeoutMs = 3000;
const initialBytesRead = 20000;

// Yup, we're single threaded. Thanks SFCC API!
const requestsMaxCount = 1;
const requestIntervalMs = 10;
let requestsPending = 0;

const axios = Axios.create();

// Thank you @matthewsuan! https://gist.github.com/matthewsuan/2bdc9e7f459d5b073d58d1ebc0613169
// Axios Request Interceptor
axios.interceptors.request.use(function (config) {
	return new Promise((resolve, reject) => {
		let interval = setInterval(() => {
			if (requestsPending < requestsMaxCount) {
				requestsPending++;
				clearInterval(interval);
				resolve(config);
			}
		}, requestIntervalMs);
	})
})

// Axios Response Interceptor
axios.interceptors.response.use(function (response) {
	requestsPending = Math.max(0, requestsPending - 1);
	return Promise.resolve(response);
}, function (error) {
	requestsPending = Math.max(0, requestsPending - 1);
	return Promise.reject(error);
})

const logfetcher = {

	errorcount: 0,
	errorlimit: 5,

	makeRequest: async function (profile: DwJson, methodStr: string, url_suffix: string, headers: Map<string, string>, debug?: boolean, securityRequest?: boolean): Promise<AxiosResponse> {
		if (!this.isUsingBM(profile) && !this.isUsingAPI(profile)) {
			this.logMissingAuthCredentials();
			process.exit(1);
		}

		let url = `https://${profile.hostname}/on/demandware.servlet/webdav/Sites/Logs`;
		let method: Method = (methodStr as Method);

		if (profile.log_source && securityRequest) {
			url = `https://${profile.hostname}/on/demandware.servlet/webdav/Sites/Securitylogs`;
		}
		if (url_suffix && url_suffix.length > 0) {
			url += '/' + url_suffix;
		}

		let opts: AxiosRequestConfig = {
			method: method,
			timeout: timeoutMs,
			url: url,
			headers: {}
		}

		if (this.isUsingBM(profile)) {
			opts.headers.Authorization = 'Basic ' + Buffer.from(profile.username + ':' + profile.password).toString('base64');
		} else {
			await this.authorize(profile, debug);
			opts.headers.Authorization = profile.token;
		}

		if (headers && headers.size > 0) {
			for (let [key, value] of headers) {
				opts.headers[key] = value;
			}
		}

		// logger.log(logger.debug, `Request: ${JSON.stringify(opts)}`, debug);
		return axios.request(opts);
	},

	authorize: async function (profile: DwJson, debug?: boolean): Promise<void> {
		if (!this.isUsingAPI(profile)) {
			this.logMissingAuthCredentials();
			process.exit(1);
		}

		if (!profile.token || !profile.token_expiry || moment.utc().isSameOrAfter(profile.token_expiry)) {
			logger.log(logger.debug, `Client API token expired or not set, resetting Client API token.`);
		} else {
			return;
		}

		let opts: AxiosRequestConfig = {
			url: 'https://account.demandware.com/dw/oauth2/access_token?grant_type=client_credentials',
			method: 'POST',
			headers: {
				'Content-Type': 'application/x-www-form-urlencoded'
			},
			auth: {
				username: profile.client_id,
				password: profile.client_secret
			}
		}
		// logger.log(logger.debug, `Request: ${JSON.stringify(opts)}`, debug);

		try {
			logger.log(logger.debug, `Authenticating to client API using client id ${profile.client_id}`, debug);
			const response = await axios.request(opts);
			profile.token = response.data.token_type.trim() + ' ' + response.data.access_token.trim();
			profile.token_expiry = moment.utc().add(response.data.expires_in, 's').subtract(profile.polling_interval, 's');
			logger.log(logger.debug, `Authenticated, token expires at ${profile.token_expiry.toString()}`, debug);
		} catch (err) {
			logger.log(logger.error, `Error authenticating client id ${profile.client_id} - please check your credentials.\n${err}.`);
			process.exit(1);
		}
	},

	fetchLogList: async function (profile: DwJson, debug?: boolean, logpath = ''): Promise<string> {
		try {
			if (!logpath || logpath.length === 0) {
				logger.log(logger.debug, `Fetching log list from ${profile.hostname}`, debug);
			} else {
				logger.log(logger.debug, `Fetching log list from ${profile.hostname}, subdirectory ${logpath}`, debug);
			}
			let headers = new Map([["User-Agent", ua]]);
			let res;
			let securityLog;
			if (profile.log_source && profile.log_source === "all") {
				securityLog = true;
				let resSite = await this.makeRequest(profile, 'GET', logpath, headers, debug);
				let resSecurity;
				let resData;
				resSecurity = await this.makeRequest(profile, 'GET', logpath, headers, debug, securityLog);
				resData = resSecurity.data.concat(resSite.data);
				res = { ...resSite, ...resSecurity };
				res.data = resData;
			}
			else if (profile.log_source && profile.log_source === "security") {
				securityLog = true;
				res = await this.makeRequest(profile, 'GET', logpath, headers, debug, securityLog);
			} else if (profile.log_source && profile.log_source === "operational") {

				res = await this.makeRequest(profile, 'GET', logpath, headers, debug);
			} else {
				res = await this.makeRequest(profile, 'GET', logpath, headers, debug);
			}
			return res.data;
		} catch (err) {
			logger.log(logger.error, 'Fetching log list failed with error: ' + err.message + logpath);
			switch (err.status) {
				case 401:
					logger.log(logger.error, 'Authentication successful but access to logs folder has been denied.');
					logger.log(logger.error, 'Please add required webdav permissions in BM -> Administration -> Organization -> WebDAV Client Permissions.');
					logger.log(logger.error, 'Sample permissions:');
					logger.log(logger.error, fs.readFileSync(path.join(__dirname, '../webdav-permissions-sample.json'), 'utf8'));
					log('\n');
					logger.log(logger.error, 'Exiting cctail.');
					process.exit(1);
				case 500:
					logger.log(logger.error, 'Authentication successful but attempt to retrieve WebDAV logs failed.');
					logger.log(logger.error, 'Please ensure your WebDAV permissions are syntactically correct and have no duplicate entries.');
					logger.log(logger.error, 'Check in BM -> Administration -> Organization -> WebDAV Client Permissions.');
					logger.log(logger.error, 'Sample permissions:');
					logger.log(logger.error, fs.readFileSync(path.join(__dirname, '../webdav-permissions-sample.json'), 'utf8'));
					log('\n');
					logger.log(logger.error, 'Exiting cctail.');
					process.exit(1);
				default:
					return '';
			}
		}
	},

	fetchFileSize: async function (profile: DwJson, logobj: LogFile): Promise<number> {
		let size = 0;
		try {
			logger.log(logger.debug, chalk.cyan(`Fetching size for ${logobj.log}`), logobj.debug,);
			let res = await this.makeRequest(profile, 'HEAD', logobj.log, null, logobj.debug, logobj.logSource === 'Securitylogs' ? true : false);
			if (res.headers['content-length']) {
				size = parseInt(res.headers['content-length'], 10);
				logger.log(logger.debug, `Fetched size for ${logobj.log}: size ${size}`, logobj.debug);
			} else {
				logger.log(logger.debug, `No content-length returned for ${logobj.log}`, logobj.debug);
			}
		} catch (err) {
			logger.log(logger.error, `Fetching file size of ${logobj.log} failed with error: ${err.message}`);
		}
		return size;
	},

	fetchLogContent: async function (profile: DwJson, logobj: LogFile): Promise<[LogFile, string]> {
		try {
			// If logobj.size is negative, leave as-is but range starts at 0. (Log rollover case)
			let range = 0;
			if (logobj.log.endsWith("log")) {
				if (!logobj.size) {
					let size = await this.fetchFileSize(profile, logobj);
					range = logobj.size = Math.max(size - initialBytesRead, 0);
				} else if (logobj.size > 0) {
					range = logobj.size;
				}
			} else {
				logobj.size = -1;
			}

			let headers = new Map([["Range", `bytes=${range}-`]]);
			let res = await this.makeRequest(profile, 'GET', logobj.log, headers, logobj.debug, logobj.logSource === 'Securitylogs' ? true : false);
			logger.log(logger.debug, `Fetching contents from ${logobj.log} source: ${logobj.logSource} retured status code ${res.status}`, logobj.debug);
			if (res.status === 206) {
				if (logobj.size < 0) {
					logobj.size = res.data.length;
					return [logobj, res.data];
				}
				if (logobj.size === 0 && res.data.length > initialBytesRead) {
					logobj.size = res.data.length;
					return [logobj, res.data.substring(res.data.length - initialBytesRead)];
				}
				logobj.size += res.data.length;
				return [logobj, res.data];
			}
		} catch (err) {
			if (err.response) {
				logger.log(logger.debug, `Fetching contents from ${logobj.log} source: ${logobj.logSource} returned status code ${err.response.status}`, logobj.debug);
			}
			if (!err.response || err.response.status !== 416) {
				this.errorcount = this.errorcount + 1;
				if (this.errorcount > 1) {
					logger.log(logger.error, `Error fetching contents from ${logobj.log} source(${logobj.logSource}): ${err.message} (error count ${this.errorcount})`);
				} else {
					// don't be too verbose, just retry if this was the first error
					logger.log(logger.debug, `Error fetching contents from ${logobj.log} source(${logobj.logSource}): ${err.message} (error count ${this.errorcount})`);
				}
			}
		}
		return [logobj, ''];
	},

	logMissingAuthCredentials: function () {
		logger.log(logger.error, ('Missing authentication credentials. Please add client_id/client_secret or username/password to log.conf.json or dw.json.'));
		logger.log(logger.error, (`Sample config:\n`));
		logger.log(logger.error, (fs.readFileSync(path.join(__dirname, '../log.config-sample.json'), 'utf8')));
		log('\n');
	},

	isUsingAPI: function (profile: DwJson) {
		return (profile.client_id && profile.client_secret)
	},

	isUsingBM: function (profile: DwJson) {
		return (profile.username && profile.password)
	}
}

export default logfetcher;


src/cctail.ts
#!/usr/bin/env node

import path from 'path';
import prompts, { Choice } from 'prompts';
import chalk from 'chalk';
import fs from 'fs';
import moment from 'moment';
import yargs from 'yargs';
import s from 'underscore.string';

import logfetcher from './lib/logfetcher';
import logparser from './lib/logparser';
import logemitter from './lib/logemitter';
import logger from './lib/logger'
import LogFluent from './lib/logfluent';
import { LogConfig, LogFile, DwJson, Profiles, FluentConfig } from './lib/types';

let fluent: LogFluent;
let logConfig: LogConfig;
let profiles: Profiles;
let profile: DwJson;
let debug = false;
let interactive = true;
let pollingSeconds = 3;
let refreshLogListSeconds = 600;
let nextLogRefresh: moment.Moment;
let latestCodeprofilerLogSent: LogFile;
let envVarPrefix = "ENV_";

let run = async function () {
	let packageJson = JSON.parse(fs.readFileSync(path.join(__dirname, '../package.json'), 'utf8'));
	logger.log(logger.info, `cctail - v${packageJson.version} - (c) openmind`);

	readLogConf();

	if (!profiles || Object.keys(profiles).length === 0) {
		logger.log(logger.warn, `No profiles in log.conf.json, checking for dw.json in path ${process.cwd()}\n`);
		readDwJson();
	}

	yargs.parserConfiguration({
		"parse-numbers": false
	});

	const args = yargs.argv

	if (args.d) {
		debug = true;
	}

	let fileobjs: LogFile[] = [];
	if (interactive) {
		fileobjs = await interact(args._[0] as string);
	} else {
		fileobjs = await dontInteract(args._[0] as string);
	}

	if (fileobjs.length === 0) {
		logger.log(logger.error, 'ERROR: No logs selected or returned, exiting.');
		process.exit(-1);
	}

	setImmediate(pollLogs, fileobjs);
}

let dontInteract = async function (profilename?: string): Promise<LogFile[]> {
	if (!profile) {
		if (Object.keys(profiles).length === 1) {
			profile = profiles[Object.keys(profiles)[0]];
		} else if (!profilename) {
			logger.log(logger.error, 'ERROR: No profile selected, exiting.');
			process.exit(-1);
		} else if (!profiles[`${profilename}`]) {
			logger.log(logger.error, `ERROR: Specified profile ${profilename} not found.`);
			process.exit(-1);
		} else {
			profile = profiles[profilename];
			logger.log(logger.info, `Using profile ${profilename}.`);
		}

		setPollingInterval(profile);
		if (profile.refresh_loglist_interval) {
			refreshLogListSeconds = profile.refresh_loglist_interval;
			logger.log(logger.info, `Setting log list refresh interval (seconds): ${refreshLogListSeconds}`);
		} else {
			profile.refresh_loglist_interval = refreshLogListSeconds;
			logger.log(logger.info, `Using default log list refresh interval (seconds): ${refreshLogListSeconds}`);
		}
	}

	nextLogRefresh = moment().add(refreshLogListSeconds, 's');
	let fileobjs = await getThatLogList(profile);
	let logx: LogFile[] = [];
	if (profile.log_types && profile.log_types.length > 0) {
		for (let thisfile of fileobjs) {
			let logname = thisfile.log.substr(0, thisfile.log.indexOf('-'));
			if (profile.log_types.indexOf(logname) != -1) {
				logx.push(thisfile);
			}
		}
	} else {
		logx = fileobjs;
	}

	if (!profile.log_types || profile.log_types.indexOf('codeprofiler') > 0) {
		let cpfileobjs = await getThatLogList(profile, '.csv');
		if (cpfileobjs && cpfileobjs.length > 0) {
			let newestcpfile = cpfileobjs.reduce((newest, compare) => newest.date.isAfter(compare.date) ? newest : compare);
			if (!latestCodeprofilerLogSent || newestcpfile.date.isAfter(latestCodeprofilerLogSent.date)) {
				logx.push(newestcpfile);
				latestCodeprofilerLogSent = newestcpfile;
			}
		}
	}

	return logx;
}

let interact = async function (profilename?: string): Promise<LogFile[]> {
	if (!profile) {
		if (Object.keys(profiles).length === 1) {
			profile = profiles[Object.keys(profiles)[0]];
		} else {
			if (profilename === undefined) {
				const profileselection = await prompts({
					type: 'select',
					name: 'value',
					message: 'Select a profile:',
					choices: Object.keys(profiles).map(i => ({
						title: `  [${i}] ${profiles[i].hostname}`,
						value: `${i}`
					}))
				});
				profilename = profileselection.value;
			}

			if (!profilename) {
				logger.log(logger.error, 'ERROR: No profile selected, exiting.');
				process.exit(-1);
			}

			if (!profiles[`${profilename}`]) {
				logger.log(logger.error, `ERROR: Specified profile ${profilename} not found.`);
				process.exit(-1);
			}

			profile = profiles[profilename];
		}
		setPollingInterval(profile);
	}

	let fileobjs = await getThatLogList(profile);
	fileobjs.sort((a, b) => b.date.unix() - a.date.unix());

	let logx: LogFile[] = [];
	let logchoiche: Choice[] = [];

	for (let i in fileobjs) {
		let sizeformatted = s.lpad(fileobjs[i].size_string, 12);
		if (sizeformatted.trim() !== '0.0 kb') {
			sizeformatted = chalk.yellow(sizeformatted);
		}
		let dateformatted = s.lpad(fileobjs[i].date.format('YYYY-MM-DD HH:mm:ss'), 20);
		if (fileobjs[i].date.isSame(moment.utc(), 'hour')) {
			dateformatted = chalk.yellow(dateformatted);
		}
		let logname = s.rpad(fileobjs[i].log, 70);

		logname = logger.colorize(logname, logname);

		logchoiche.push({
			title: `${chalk.green(s.lpad(i, 2))} ${logname} ${sizeformatted}  ${dateformatted}`,
			value: i
		});
	}

	let logselection = await prompts({
		type: 'autocompleteMultiselect',
		name: 'value',
		message: `Select logs on [${chalk.green(profile.hostname)}]`,
		choices: logchoiche,
		// eslint-disable-next-line no-return-assign
		onState: ((statedata) => { statedata.value.forEach((i: Choice) => i.title = `\n${i.title}`) })
	});

	if (logselection.value) { // ctrl+c
		logselection.value.forEach((i: number) => {
			logx.push(fileobjs[i]);
		});
	}

	return logx;
}

let setPollingInterval = function (profile: DwJson) {
	if (profile.polling_interval) {
		pollingSeconds = profile.polling_interval;
		logger.log(logger.info, `Setting polling interval (seconds): ${pollingSeconds}`);
	} else {
		logger.log(logger.info, `Using default polling interval (seconds): ${pollingSeconds}`);
		profile.polling_interval = pollingSeconds;
	}
}


let getThatLogList = async function (profile: DwJson, filesuffix = ".log"): Promise<LogFile[]> {
	let fileobjs: LogFile[] = [];

	let data = '';

	if (filesuffix === ".csv") {
		data = await logfetcher.fetchLogList(profile, debug, 'codeprofiler');
	} else {
		data = await logfetcher.fetchLogList(profile, debug);
	}
	let regexp = new RegExp(`<a href="/on/demandware.servlet/webdav/Sites/(.*?)/(.*?)">[\\s\\S\\&\\?]*?<td align="right">(?:<tt>)?(.*?)(?:<\\/tt>)?</td>[\\s\\S\\&\\?]*?<td align="right"><tt>(.*?)</tt></td>`, 'gim');
	let match = regexp.exec(data);
	while (match != null) {
		let filedate = moment.utc(match[4]);
		if (match[2].substr(-4) === filesuffix && filedate.isSame(moment.utc(), 'day')) {
			fileobjs.push({
				log: match[2],
				size_string: match[3],
				date: moment.utc(match[4]),
				debug: debug,
				logSource: match[1]
			});
			logger.log(logger.debug, `Available Log: ${match[2]},log size: ${match[3]}, log source: ${match[1]}`, debug);
		}
		match = regexp.exec(data);
	}

	return fileobjs;
}

let pollLogs = async function (fileobjs: LogFile[], doRollover = false) {
	if (logfetcher.isUsingAPI(profile) && logfetcher.errorcount > logfetcher.errorlimit) {
		logger.log(logger.error, `Error count (${logfetcher.errorcount}) exceeded limit of ${logfetcher.errorlimit}, resetting Client API token.`);
		logfetcher.errorcount = 0;
		profile.token = null;
		await logfetcher.authorize(profile, debug);
	}

	if (!doRollover) {
		if (moment.utc().isAfter(fileobjs[0].date, 'day')) {
			logger.log(logger.info, 'Logs have rolled over, collecting last entries from old logs.');
			doRollover = true;
		} else {
			logger.log(logger.debug, 'Logs have not rolled over since last poll cycle.', debug);
			if (nextLogRefresh && moment().isSameOrAfter(nextLogRefresh)) {
				logger.log(logger.debug, 'Refreshing log list.', debug);
				let newfiles = await dontInteract();
				for (let newfile of newfiles) {
					if (!fileobjs.some(logfile => logfile.log === newfile.log)) {
						logger.log(logger.debug, `Added new log file: ${newfile.log}.`, debug);
						fileobjs.push(newfile);
					}
				}
			}
		}

		if (fluent) {
			fluent.output(profile.hostname,
				await logparser.process(fileobjs.map((logobj) => logfetcher.fetchLogContent(profile, logobj))),
				false, fileobjs[0].debug);
		} else {
			let parsed = logemitter.sort(
				await logparser.process(fileobjs.map((logobj) => logfetcher.fetchLogContent(profile, logobj)))
			);
			logemitter.output(parsed, false, fileobjs[0].debug);
		}

		// Codeprofiler files should only be consumed once
		let cp = fileobjs.findIndex(logobj => logobj.log.endsWith("csv"));
		if (cp > -1) {
			logger.log(logger.debug, `Removed codeprofiler log ${fileobjs[cp].log} from list.`, debug);
			fileobjs.splice(cp, 1);
		}

	} else {
		if (interactive) {
			fileobjs = await interact();
		} else {
			fileobjs = await dontInteract();
		}

		if (fileobjs.length != 0) {
			doRollover = false;
			for (let i of fileobjs) {
				i.size = -1;
			}
		} else {
			logger.log(logger.warn, 'No logs to report yet, waiting until next cycle.');
		}
	}

	setTimeout(pollLogs, pollingSeconds * 1000, fileobjs, doRollover);
}

function replaceEnvPlaceholders(data: any) {
	Object.keys(data).forEach(function (key) {
		var value = data[key];
		if (typeof (value) === 'object') {
			replaceEnvPlaceholders(value);
		} else if (typeof (value) === 'string' && value.startsWith(envVarPrefix)) {
			var checkForVar = value.replace(envVarPrefix, "");
			if (process.env.hasOwnProperty(checkForVar)) {
				data[key] = process.env[checkForVar];
			}
		}
	});
	return data;
}

function readDwJson() {
	let dwJsonPath = path.join(process.cwd(), 'dw.json');
	logger.log(logger.info, `Loading profile from ${dwJsonPath}\n`);
	try {
		const dwJson = replaceEnvPlaceholders(JSON.parse(fs.readFileSync(dwJsonPath, 'utf8')));
		const name = dwJson.profile || dwJson.hostname.split('-')[0].split('-')[0];
		profiles[name] = dwJson;
	}
	catch (err) {
		logger.log(logger.error, `No dw.json found in path ${process.cwd()}\n`);
		process.exit(-1);
	}
}

function readLogConf() {
	try {
		logConfig = replaceEnvPlaceholders(JSON.parse(fs.readFileSync(`${process.cwd()}/log.conf.json`, 'utf8')));
		profiles = logConfig.profiles ?? logConfig as any; // support for old configs (without "profiles" group)
		if (logConfig.interactive !== undefined && logConfig.interactive === false) {
			interactive = false;
			logger.log(logger.info, "Interactive mode is disabled.");
		}
		if (logConfig.fluent && logConfig.fluent.enabled) {
			let fluentConfig: FluentConfig = logConfig.fluent;
			fluent = new LogFluent(fluentConfig);
			logger.log(logger.info, "FluentD output is enabled.");
		} else {
			logger.log(logger.info, "Console output is enabled.");
		}
	} catch (err) {
		logger.log(logger.error, `\nMissing or invalid log.conf.json.\nError message: ${err}\n`);
		process.exit(-1);
	}
}

run();
